src/librustc_codegen_ssa/mir/block.rs

```diff
@@ -548,28 +528,33 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
             _ => FnType::new(&bx, sig, &extra_args)
         };
 
+        if intrinsic == Some("caller_location") {
+            if let Some((dest, target)) = destination.as_ref() {
+                let loc = bx.sess().source_map().lookup_char_pos(span.lo());
+                let location = bx.static_panic_location(&loc);
+
+                let mut llargs = vec![];
+                let dest = self.make_return_dest(&mut bx, dest, &fn_ty.ret, &mut llargs, true);
+                self.store_return(&mut bx, dest, &fn_ty.ret, location);
+                helper.maybe_sideeffect(self.mir, &mut bx, &[*target]);
+                helper.funclet_br(self, &mut bx, *target);
+            }
+            return;
+        }
+
         // Emit a panic or a no-op for `panic_if_uninhabited`.
         if intrinsic == Some("panic_if_uninhabited") {
             let ty = instance.unwrap().substs.type_at(0);
             let layout = bx.layout_of(ty);
             if layout.abi.is_uninhabited() {
                 let loc = bx.sess().source_map().lookup_char_pos(span.lo());
-                let filename = Symbol::intern(&loc.file.name.to_string());
-                let line = bx.const_u32(loc.line as u32);
-                let col = bx.const_u32(loc.col.to_usize() as u32 + 1);
+                let location = bx.static_panic_location(&loc);
 
                 let str = format!(
                     "Attempted to instantiate uninhabited type {}",
```

Comment:
Drive-by style nit (for @RalfJung?): messages should start with lowercase and types should have backticks around them, i.e.:
```rust
"attempted to instantiate uninhabited type `{}`"
```